server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java

```diff
@@ -344,4 +437,90 @@ private static Map<String, Object> encodeRecord(Map<String, ColumnType> columnTy
         }
         return ret;
     }
+
+    private boolean saveOneTable(long minNoUpdatePeriodMillis) {
+        for (var table : this.dirtyTables.keySet()) {
+            table.lock();
+            try {
+                var now = System.currentTimeMillis();
+                if ((now - table.getLastUpdateTime()) >= minNoUpdatePeriodMillis) {
+                    table.save();
+                    this.dirtyTables.remove(table);
+                    return true;
+                }
+            } finally {
+                table.unlock();
+            }
+        }
+        return false;
+    }
+
+    private void clearWalLogFiles() {
+        // should be put after getting the table list, so that
+        long minWalLogIdToRetain = this.walManager.getMaxEntryId() + 1;
+        for (var table : this.dirtyTables.keySet()) {
+            table.lock();
+            try {
+                if (table.getFirstWalLogId() < minWalLogIdToRetain) {
+                    minWalLogIdToRetain = table.getFirstWalLogId();
+                }
+            } finally {
+                table.unlock();
+            }
+        }
+        try {
+            this.walManager.removeWalLogFiles(minWalLogIdToRetain);
```

Comment:
I guess we should record the global max entry id, this fucntion may remove all of the wal files.
Or always leave the last wal file

Reply:
If there is any dirty table, whose firstWalLogId should be greater than 0 and less than the latest walLogId, the corresponding WAL files will not be removed.

Reply:
Fixed